[fix] solidtorrents engine - JSON API no longer exists

return42 opened the pull request:
The API endpoint, we where using does not exist anymore.  This patch is a
rewrite that parses the HTML page.

Related: https://github.com/paulgoio/searxng/issues/17
Closes: https://github.com/searxng/searxng/issues/858

return42 committed:
[fix] solidtorrents engine: JSON API no longer exists

The API endpoint, we where using does not exist anymore.  This patch is a
rewrite that parses the HTML page.

Related: https://github.com/paulgoio/searxng/issues/17
Closes: https://github.com/searxng/searxng/issues/858

Signed-off-by: Markus Heiser <markus.heiser@darmarit.de>

dalf commented on searx/engines/solidtorrents.py:
```diff
@@ -1,51 +1,85 @@
+base_url = ''
+base_url_rand = ''
+
+units = {"B": 1, "KB": 2 ** 10, "MB": 2 ** 20, "GB": 2 ** 30, "TB": 2 ** 40}
```
Note: [`get_torrent_size`](https://github.com/searxng/searxng/blob/de32d543bc5e7fac968a11c7cdd0e3b8f349a94d/searx/utils.py#L263-L301) in `searx.utils` convert the file size to bytes. The current implementation of this function misses the `B` unit.

The PR https://github.com/searxng/searxng/pull/839 improves the implementation: https://github.com/searxng/searxng/blob/2d5929cc59f30ab4c151c69d797da64c87df2a0a/searx/utils.py#L41-L48

return42 replied:
Thanks for pointing out [get_torrent_size](https://github.com/searxng/searxng/blob/de32d543bc5e7fac968a11c7cdd0e3b8f349a94d/searx/utils.py#L263-L301) I haven't seen / fixed.

dalf commented on searx/engines/solidtorrents.py:
```diff
@@ -1,51 +1,85 @@
+            'seed': seed,
+            'leech': leech,
+            'title': title,
+            'url': base_url_rand + url,
```
Note without consequence / can be ignored: the `base_url` here might be different the `base_url` in the `request` function because two or more user query can run in parallel.

To keep exactly the same value `base_url` must be stored in `params` and then retrieve using `resp.search_params["base_url"]`.

return42 replied:
> To keep exactly the same value `base_url` must be stored in `params` and then retrieve using `resp.search_params["base_url"]`.

AFAIK request & response are in the same process (task) environment .. isn't it?

I copied this method from

https://github.com/searxng/searxng/blob/de32d543bc5e7fac968a11c7cdd0e3b8f349a94d/searx/engines/invidious.py#L43-L49

https://github.com/searxng/searxng/blob/de32d543bc5e7fac968a11c7cdd0e3b8f349a94d/searx/engines/invidious.py#L67-L73

dalf replied:
Two different threads ( = two different user queries) can call the `request` function in a row and then the `response` function. The namespace will be same since this is the same engine.

return42 replied:
:facepalm: thanks for clarify, I fixed it here in solidtorrents engine and in invidious engine https://github.com/searxng/searxng/pull/867

return42 committed:
[fix] solidtorrents engine: use get_torrent_size from searx.utils

Suggested-by: @dalf https://github.com/searxng/searxng/pull/862#pullrequestreview-872858489
Signed-off-by: Markus Heiser <markus.heiser@darmarit.de>

dalf commented on searx/engines/solidtorrents.py:
```diff
@@ -58,7 +56,8 @@ def response(resp):
         title = extract_text(a)
         url = eval_xpath_getindex(a, '@href', 0, None)
         stats = eval_xpath(result, './div//div[contains(@class, "stats")]/div')
```
Nit: it is possible to write `stats = eval_xpath_list(result, './div//div[contains(@class, "stats")]/div', min_len=5)` so when the HTML changes, instead of crashing with an `IndexError` few lines later, it crashes with `searx.exceptions.SearxEngineXPathException: ./div//div[contains(@class, "stats")]/div len(xpath_str) < 5`.

return42 replied:
good point / I fixed it to use min_len

@dalf Thanks a lot for your pragmatic suggestions & comments! .. I'm going to merge this PR ..

dalf reviewed (approved):
Thank you!

return42 committed:
[mod] solidtorrents engine: add metadata & torrentfile

BTW: define min_len in eval_xpath_list of 'stats' list

Suggested-by: @dalf https://github.com/searxng/searxng/pull/862#pullrequestreview-872910744
Signed-off-by: Markus Heiser <markus.heiser@darmarit.de>

return42 committed:
[fix] solidtorrents engine: store random bas_url in param

Two different threads ( = two different user queries) can call the request
function in a row and then the response function.  The namespace will be same
since this is the same engine.

To keep exactly the same value ``base_url`` must be stored in params and then
retrieve using ``resp.search_params["base_url"]``.

Suggested-by: @dalf https://github.com/searxng/searxng/pull/862#discussion_r799324861
Signed-off-by: Markus Heiser <markus.heiser@darmarit.de>
